Show Active Connections

danieldresser-ie opened the pull request:
Still lots to do, but I thought I would upload this in case you wanted to comment on the basic framework ( it's just one commit, but is built on top of the focus nodes PR )

Still some UI niceties to improve, but I think this skeleton should work.

Still needing to be done:

* UI tweaks
* trigger update whenever anything that could affect activations changes ( not just changing focus )
* add cancellation support to activation walk
* make activation walk actually track contexts and understand switches, instead of naively taking all inputs all the time

Is `setActive` on NodeGadget the right name?  "active" is the name Tom used in his description, but it sounds to me like it could be something that affects more than just being a UI hint related to the focus node.

danieldresser-ie commented:
Seems like the logic is now basically all working, including switches, context processors, loops, and even the same switch being evaluated in two different contexts.

The hardest remaining part is probably aesthetics.  The last commit switches to a completely different look where bright outlines are used to show the active part - it's pretty garish at the moment, but maybe it suggests an idea to you.  If you revert that, you get the dimming based version, which is a bit less clear, but currently a lot less ugly.

danieldresser-ie commented:
Hmm, just saw a weird deadlock when the background task used by this PR tries to call `callOnUIThread` at the same time as `GafferScene::InteractiveRender::messagesChangedUI()` does a plug set on the UI thread which calls `Gaffer::BackgroundTask::cancelAffectedTasks`.  That needs fixing.

danieldresser-ie commented:
Did some work on this.  Currently your feedback is in the Gaffer script you provided, but I'll summarize where things are at here ( the code is in a bit of a mess right now, there will be a refactor needed once we get everything working how we want ).

Focussing directly on NameSwitch: was due to activeInPlug() returning the NameValuePlug, not the value plug that actually gets connected.  Currently special cased, may require addressing some other way.  Raises general question about handling parent/child plugs.

Box inputs : Addressed

Disabled nodes activating correspondingInput only:  Addressed

Selected nodes get highlighted : Due to quick hack where I currently use the "State" argument set to disabled in Style::renderConnection to specify that a node is inactive.  Clearly, a node can be highlighted or unhighlighted while being active or inactive, so this is not a good API.  I may need to change the API on Style to address this properly.

Grey out Strikethrough for disabled nodes : Addressed, though I'm not sure if it looks better

Dimming vs highlighting:  We should be able to get a pure dimming approach working, but it will mean a bunch of time spent on selecting nodule/noodle colors and trying to get IE to synchronize on our colors.  I'm still sort of feeling like highlighting connections could be better, so I've taken one last swing at styling connection highlighting - now rendering it more as an actual glow, with a fixed size in pixels ( within a min/max ).  This is actually looking like it might be getting close to somewhere to me, but I'm OK to focus on dimming instead if you think this is a dead end.

What to do when no focus : Pending figuring out dimming vs highlighting

Unit tests : Not yet Addressed
Wedge Contexts : Not yet addressed
Cancellation: Not yet addressed

danieldresser-ie commented:
Talking to Andrew and Murray, there was some suggestion that the highlighted version looked pretty decent up close, but wasn't working at a further zooms.  I had a go a playing with tuning how it looks from farther out - probably not quite there yet, but might be worth you having a look.

I'm also thinking maybe we need to increase the contrast between active and inactive nodes, though that will hurt legibility of node colors of inactive nodes.

Also, added a hack special case for the index plug on Switch and NameSwitch.

johnhaddon commented:
> Talking to Andrew and Murray, there was some suggestion that the highlighted version looked pretty decent up close, but wasn't working at a further zooms. I had a go a playing with tuning how it looks from farther out - probably not quite there yet, but might be worth you having a look.

This feels like a big improvement on the zoomed-out look of the previous version. I'm still pretty unsure whether I would personally push for highlighting vs dimming, but I think this convinces me that I could get behind highlighting if that's the general consensus.

> I'm also thinking maybe we need to increase the contrast between active and inactive nodes, though that will hurt legibility of node colors of inactive nodes.

I was wondering the opposite! If the connection highlighting reads well enough on its own, perhaps we don't even need dimming, and can preserve that visual bandwidth for communicating other stuff...

danieldresser-ie commented:
> I was wondering the opposite! If the connection highlighting reads well enough on its own, perhaps we don't even need dimming, and can preserve that visual bandwidth for communicating other stuff...

Hmm, that is an idea.  I feel like the current node dimming isn't communicating much, so yeah, the options would be dim more, don't indicate active nodes at all ( just connections ), or try out some subtle highlighting on nodes as well.

I'll try to gather some more user feedback - I'm also feeling like I've fiddled with the style enough for the moment, and I should go back and finish off the issues with the actual functionality that you identified.

danieldresser-ie commented:
This code is currently in a giant mess, but it does now allow some temporary playing around with visual styling.

The temporary reconfiguration of the visual style is done with:
`GafferUI.StandardStyle.setGlowPrototypeAppearance( imath.Color4f( 1, 1, 1, 0.2 ), 1.0, True )`
( Those are my current default values )

You suggested maybe disabling the node glow, which could be tested with:
`GafferUI.StandardStyle.setGlowPrototypeAppearance( imath.Color4f( 1, 1, 1, 0.2 ), 1.0, False )`
You may want to raise the brightness to compensate for being less visible, like this:
`GafferUI.StandardStyle.setGlowPrototypeAppearance( imath.Color4f( 1, 1, 1, 0.5 ), 1.0, False )`

In my one casual user test so far, I may have influenced the user, but they preferred with node glow, but a little bit brighter than my default, something like:
`GafferUI.StandardStyle.setGlowPrototypeAppearance( imath.Color4f( 1, 1, 1, 0.3 ), 1.0, True )`

The final commit currently was a hasty attempt to switch over to triggering updateActive from renderLayer, which I think you suggested.  This fixes an issue with cancellation, though it was last minute, so I can't rule out the possibility of having introduced other issues.

danieldresser-ie commented:
Getting closer to ready for review.  Main outstanding issues:

* still need to expose activationWalk for testing ( we're now thinking this API will be very temporary, because we're about to throw it out and replace it with something that is more flexible for adding other features onto once I finish this commit? )
* 5ae580a seems like a definite visual improvement, but the current implementation is a massive hack
* I added support for handing Wedges properly, which is nice, but this introduces a dependency on GafferDispatch which I assume is unacceptable.  Do I need to add some sort of registration mechanism to deal with this?

danieldresser-ie commented:
Oh, also:  question:  should we be dimming nodules too?  Or is it nice to clearly see the types of nodules on inactive nodes?

johnhaddon commented:
My take on the various open questions :

> Is setActive on NodeGadget the right name?

Probably not. Ultimately I'm not even sure there will be a public method for this (NodeGadget might just become a client of the `ThingWotTellsYouAboutContextsBasedOnFocus`). Maybe `activeForFocusNode( bool active )`, without a getter? Maybe we even make it private and give GraphGadget friend access for now?

> still need to expose activationWalk for testing ( we're now thinking this API will be very temporary, because we're about to throw it out and replace it with something that is more flexible for adding other features onto once I finish this commit? )

Sounds about right to me. Let's bind it with an `_` prefix to make it clear this isn't for general consumption.

> I added support for handing Wedges properly, which is nice, but this introduces a dependency on GafferDispatch which I assume is unacceptable. Do I need to add some sort of registration mechanism to deal with this?

It does feel like we'd need a registration mechanism, or possibly a new virtual method on DependencyNode. But given that we think we're going to shake up the implementation soon anyway, I suggest we hold off for now. We can live without extended Wedge support in the initial version - the fast majority of the benefit is captured in the support for the other node types. Keep that commit squirreled away though - it'd be great to bring it back soon.

> should we be dimming nodules too? Or is it nice to clearly see the types of nodules on inactive nodes?

I don't feel strongly about this, but if I had to jump off the fence right this minute, I would stick with them as they are. There's a slight danger that dimming reads as "not available for interaction" in the way that we grey out read-only widgets. I think keeping the nodules at their usual bright colour is perhaps a small but useful visual clue that they're still open for business.

I've got a question of my own about connection colours. When there is no focus node, should we be highlighting all connections? I know we conceptualised this as "we're making the default connection colours bright, so we have space to dim them into", but in practice it feels like they are actively being highlighted. When there's no focus node, it feels a little overwhelming.

johnhaddon commented:
Haven't reviewed code at all since it sounds like it's not quite ready yet, but had a quick play with it again. Which lead me to making an entry for the Node Graph Colours Bikeshedding Contest of 2021 : https://github.com/johnhaddon/gaffer/tree/theGreatBikeSheddingOf2021. I think the second commit might be questionable, but I think I do prefer the first - keeping some colour in the inactive connections helps readability a bit I think, and means that nodes/connections feel like they're doing the same thing when dimmed. See what you think...

johnhaddon commented:
> I think the second commit might be questionable,

Actually, I might take that back. It's a bit less logical (why would nodes display as active, but not connections?), but I just went back to Gaffer with fresh eyes after doing some other stuff, and I think it works quite well in practice. The "no focus yet" node graph is less in your face, and when you focus something for the first time, it feels like the dataflow in the active connections is springing to life, rather than the other connections stopping doing something.

Oh, and this might be the same as the box thing you're already looking at, but I thought I'd mention it. In the script below, if you focus ShaderAssignment1, everything highlights nicely upstream of RenderBox. But if you focus RenderBox itself, no connections are highlighted. I think this is because the scene output is actually the second output (even though its the first visually). But it makes me wonder if we should start our walk from _all_ outputs, not just the first? In my example, the HierarchyView is happily showing the scene output, even though the image output is the first one.

```
import Gaffer
import GafferArnold
import GafferDispatch
import GafferImage
import GafferScene
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 61, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["ContextVariables"] = Gaffer.ContextVariables( "ContextVariables" )
parent.addChild( __children["ContextVariables"] )
__children["ContextVariables"].setup( GafferScene.ScenePlug( "in", ) )
__children["ContextVariables"]["variables"].addChild( Gaffer.NameValuePlug( "", Gaffer.StringPlug( "value", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member1", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["ContextVariables"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"] = Gaffer.Box( "RenderBox" )
parent.addChild( __children["RenderBox"] )
__children["RenderBox"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"].addChild( GafferArnold.ArnoldRender( "ArnoldRender" ) )
__children["RenderBox"]["ArnoldRender"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"].addChild( Gaffer.BoxIn( "BoxIn" ) )
__children["RenderBox"]["BoxIn"].setup( GafferScene.ScenePlug( "out", ) )
__children["RenderBox"]["BoxIn"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"].addChild( GafferScene.ScenePlug( "in", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"].addChild( GafferImage.ImageReader( "ImageReader" ) )
__children["RenderBox"]["ImageReader"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"].addChild( Gaffer.BoxOut( "BoxOut" ) )
__children["RenderBox"]["BoxOut"].setup( GafferImage.ImagePlug( "in", ) )
__children["RenderBox"]["BoxOut"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"].addChild( GafferImage.ImagePlug( "out", direction = Gaffer.Plug.Direction.Out, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"].addChild( Gaffer.BoxOut( "BoxOut1" ) )
__children["RenderBox"]["BoxOut1"].setup( GafferScene.ScenePlug( "in", ) )
__children["RenderBox"]["BoxOut1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RenderBox"].addChild( GafferScene.ScenePlug( "out1", direction = Gaffer.Plug.Direction.Out, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShaderAssignment1"] = GafferScene.ShaderAssignment( "ShaderAssignment1" )
parent.addChild( __children["ShaderAssignment1"] )
__children["ShaderAssignment1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Blur1"] = GafferImage.Blur( "Blur1" )
parent.addChild( __children["Blur1"] )
__children["Blur1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ContextVariables"]["variables"]["member1"]["name"].setValue( 'shot' )
__children["ContextVariables"]["variables"]["member1"]["value"].setValue( 'A' )
Gaffer.Metadata.registerValue( __children["ContextVariables"]["in"], 'noduleLayout:section', 'top' )
Gaffer.Metadata.registerValue( __children["ContextVariables"]["out"], 'noduleLayout:section', 'bottom' )
__children["ContextVariables"]["__uiPosition"].setValue( imath.V2f( 27.0588074, -26.3209877 ) )
__children["RenderBox"]["__uiPosition"].setValue( imath.V2f( 50.8546677, -58.7992172 ) )
__children["RenderBox"]["ArnoldRender"]["in"].setInput( __children["RenderBox"]["BoxIn"]["out"] )
__children["RenderBox"]["ArnoldRender"]["fileName"].setValue( '${project:rootDirectory}/asses/${script:name}/${script:name}.####.ass' )
__children["RenderBox"]["ArnoldRender"]["__uiPosition"].setValue( imath.V2f( -46.1999969, -7.59999943 ) )
__children["RenderBox"]["BoxIn"]["__in"].setInput( __children["RenderBox"]["in"] )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxIn"]["__in"], 'nodule:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxIn"]["__in"], 'connectionGadget:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxIn"]["__in"], 'layout:section', 'Settings' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxIn"]["__in"], 'nodule:type', 'GafferUI::StandardNodule' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxIn"]["__in"], 'description', 'The scene to be rendered.' )
__children["RenderBox"]["BoxIn"]["__uiPosition"].setValue( imath.V2f( -33.2992821, 4.23203135 ) )
__children["RenderBox"]["in"].setInput( __children["ContextVariables"]["out"] )
Gaffer.Metadata.registerValue( __children["RenderBox"]["in"], 'nodule:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["RenderBox"]["in"], 'nodule:type', 'GafferUI::StandardNodule' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["in"], 'description', 'The scene to be rendered.' )
__children["RenderBox"]["ImageReader"]["__uiPosition"].setValue( imath.V2f( -24.9992867, -11.4000006 ) )
__children["RenderBox"]["BoxOut"]["in"].setInput( __children["RenderBox"]["ImageReader"]["out"] )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxOut"]["__out"], 'layout:section', 'Settings' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxOut"]["__out"], 'description', 'The output image generated by this node.' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxOut"]["__out"], 'nodule:type', 'GafferUI::StandardNodule' )
__children["RenderBox"]["BoxOut"]["__uiPosition"].setValue( imath.V2f( -23.4992867, -24.7320328 ) )
__children["RenderBox"]["out"].setInput( __children["RenderBox"]["BoxOut"]["__out"] )
Gaffer.Metadata.registerValue( __children["RenderBox"]["out"], 'description', 'The output image generated by this node.' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["out"], 'nodule:type', 'GafferUI::StandardNodule' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["out"], 'noduleLayout:index', 1 )
__children["RenderBox"]["BoxOut1"]["name"].setValue( 'out1' )
__children["RenderBox"]["BoxOut1"]["in"].setInput( __children["RenderBox"]["BoxIn"]["out"] )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxOut1"]["__out"], 'nodule:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxOut1"]["__out"], 'connectionGadget:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxOut1"]["__out"], 'layout:section', 'Settings' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["BoxOut1"]["__out"], 'plugValueWidget:type', '' )
__children["RenderBox"]["BoxOut1"]["__uiPosition"].setValue( imath.V2f( -31.7992821, -24.8679695 ) )
__children["RenderBox"]["out1"].setInput( __children["RenderBox"]["BoxOut1"]["__out"] )
Gaffer.Metadata.registerValue( __children["RenderBox"]["out1"], 'nodule:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["RenderBox"]["out1"], 'connectionGadget:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["RenderBox"]["out1"], 'plugValueWidget:type', '' )
Gaffer.Metadata.registerValue( __children["RenderBox"]["out1"], 'noduleLayout:index', 0 )
__children["ShaderAssignment1"]["in"].setInput( __children["RenderBox"]["out1"] )
__children["ShaderAssignment1"]["__uiPosition"].setValue( imath.V2f( 36.4472122, -74.9016418 ) )
__children["Blur1"]["in"].setInput( __children["RenderBox"]["out"] )
__children["Blur1"]["radius"]["y"].setInput( __children["Blur1"]["radius"]["x"] )
__children["Blur1"]["__uiPosition"].setValue( imath.V2f( 57.7758865, -74.6191559 ) )


del __children

```

johnhaddon commented:
Just noticed we have a CI crash when generating the docs :

```* CRASHED in (anonymous namespace)::activationWalk(Gaffer::Plug const*, Gaffer::Context const*, IECore::Canceller const*, std::unordered_set<Gaffer::Plug const*, std::hash<Gaffer::Plug const*>, std::equal_to<Gaffer::Plug const*>, std::allocator<Gaffer::Plug const*> >&, std::unordered_set<Gaffer::Node const*, std::hash<Gaffer::Node const*>, std::equal_to<Gaffer::Node const*>, std::allocator<Gaffer::Node const*> >&, boost::unordered::unordered_set<IECore::MurmurHash, boost::hash<IECore::MurmurHash>, std::equal_to<IECore::MurmurHash>, std::allocator<IECore::MurmurHash> >&) at 00:00:00
* signal caught: SIGSEGV -- Invalid memory reference (address not mapped to object)
*
```

Just a guess, but perhaps `~GraphGadget` needs to `cancelAndWait()` the background task?

danieldresser-ie commented:
OK, I think I've addressed all comments, and the docs seem to build OK now ( for some reason, the IE config is no longer finding Sphinx now, haven't debugged that yet, but the CI docs built OK with the last version of this PR ).

UI style wise, I think I'm OK with this.  I've tried out not desaturating the noodle colors, which I think does make things a little clearer?  Still need to survey folks at IE.  As for the behaviour of deactivating connections but not nodes when there is no focus node, I don't mind it visually, but the API seems pretty counterintuitive - it's not just a UI style thing, but we actually call activeForFocusNode( False ) for one and True for the other when there is no focus node.  That's weird, but it's a private API, and we're probably about to rework it, so I guess we don't care that it's bad?

I was trying to finish off the final commit where I rework the actual traversal behaviour to be more correct for things like expressions on the variable name of a ContextVariables node, but for some reason, that causes crashes.  I probably do need to figure out what's going on there before we merge this.

johnhaddon commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
+	ConstGraphGadgetPtr thisRef = this;
+
+	m_activeStateTask = Gaffer::ParallelAlgo::callOnBackgroundThread(
+		// Subject:  TODO - does it matter that we're only including one plug as a subject?  We respond to any
```
It definitely doesn't matter right now, because cancellation for graph edits is done at the level of the ScriptNode (any edit anywhere in the script will cancel us). I don't know if we'll ever get to doing that at a finer level of granularity, but it does seem that `callOnBackgroundThread()` needs a more flexible way of specifying a subject (or subjects). See related comments in GraphComponentPath in my recent PR.

Not something we should worry about now, but it does add to the evidence that the signature for `callOnBackgroundThread()` isn't quite right.

danieldresser-ie replied:
Clarified comment a bit in 078a377

johnhaddon commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
+			}
+
+			{
+				tbb::spin_mutex::scoped_lock scopedLock( pendingActiveStateMutex );
+				pendingActiveState.swap( newActive );
```
I've been trying to figure out why we need this `m_pendingActiveState` stuff, and I'm coming up blank. Can we just pass `newActive` directly as an argument to `applyActive()`?

danieldresser-ie replied:
It's definitely true that after several iterations, this ended up overcomplicated.

I tried a naive implementation, and it works, but my debug output is showing that I'm actually evaluating it several times ... keeping track of the callOnUIThread being done by the background thread feels like a lot of things happening on different threads.  I guess with the current logic, the background thread has finishes, and launches a callOnUIThread, but then renderLayer comes along and notices that the active state is dirty, and that something should be computing it, but there is no background thread, because it's already finished, so it launches a new background thread.  I guess I can fix this with a new flag for "waiting on a callOnUIThread"?

Perhaps a better way of simplifying would be to keep pendingActiveState, but drop the callOnUIThread from the background thread, and instead, when the foreground thread hits renderLayer, it checks if a pending activation has been set before thinking about starting a background thread?  I think I must have originally been thinking of something like this, and I'm not sure why I didn't do it - maybe it felt wrong to stall in renderLayer while transferring the activate state to all the gadgets?

danieldresser-ie replied:
It's not quite that simple - at the very least, we would need to catch the situation where m_backgroundTask still exists but is cancelled.  But maybe an approach like that could work, as long as we can absolutely depend on callOnUIThread being 100% reliable.

danieldresser-ie replied:
Logic simplified quite a bit in 078a377

johnhaddon commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
+			// and activePlugsAndNodes needs to continue even in cases where it can't figure out an accurate
+			// context andt just uses null ... but we do need to take the canceller from the context
+			// provided by callOnBackgroundThread
+			const IECore::Canceller *canceller = Gaffer::Context::current()->canceller();
```
Where do we transfer the canceller into the contexts used to evaluate the graph? Assuming we're not doing it yet, perhaps the `EditableScope::setCanceller()` method I just added might be useful? (https://github.com/GafferHQ/gaffer/pull/4368/commits/5de5ccdfb2b77c453e0c981a8325862be5bc535e)

danieldresser-ie replied:
Now that I've removed the canceller argument from the API for activePlugsAndNodes, and activeWalk is an anonymous internal function, there's no way to get here without the canceller already being in the context used to evaluate the graph.

While thinking about this further though, I realized that I haven't really confirmed that a useful Canceller is actually in the Context that updateActive gets called with.  Upon testing, I don't see any sign that cancellation is actually doing anything.  So that probably needs looking into, though I'm trying to ignore that for now and just get everything cleaned up.

danieldresser-ie replied:
I did poke a bit further, and it's now working fine.  There were two issues:  I had accidentally called `new Context( context, otherContext->canceller() )`, not realizing that without dereferencing the Canceller pointer, it gets converted to bool or something.  The other thing I ran into was that my test graph used an OSL expression to connect things, and after fixing that, I got a crash, which is fixed by: 804729b

danieldresser-ie replied:
Canceller passing is updated in a9b4961a2

johnhaddon commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
+	std::unique_ptr<PendingActive> &pendingActiveState = m_pendingActiveState;
+
+	// Must hold a reference to stop us dying before our UI thread call is scheduled.
+	ConstGraphGadgetPtr thisRef = this;
```
I think we're on modern enough C++ that we can do this in the lambda capture itself rather than needing to make a variable. Something like `[ thisRef = ConstGraphGadgetPtr( this ) ]`.

danieldresser-ie replied:
Addressed in 2ddd2b2e

johnhaddon commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
+
+	m_activeStateDirty = false;
+
+	// TODO - this const_cast is pretty ugly
```
Seems like `updateActive()` and `applyActive()` should be non-const since they do mutate stuff, and the source of the problem is `updateActive()` being called from `doRenderLayer()`? Maybe the ugly cast belongs there, perhaps with a comment about us being in trouble if ViewportGadget changes drawing strategy to include display lists? I dunno, given that we're thinking we might overhaul the activation mechanism shortly anyway, maybe it's not worth worrying.

danieldresser-ie replied:
Addressed in 2ddd2b2e

I'm not sure a display list drawing strategy specifically is the problem - we would still need to rebuild the display lists when the graph is edited, and rebuilding the display list could trigger the activation walk.  It's maybe a little odd to wait for the display list rebuild before starting to think about calculating active, and needing another display list rebuild when it finishes, but the same is basically true of the current code if you're not panning around to require rerendering.

johnhaddon commented on src/GafferUI/TextGadget.cpp:
```diff
@@ -90,7 +99,15 @@ void TextGadget::renderLayer( Layer layer, const Style *style, RenderReason reas
 	}
 
-	style->renderText( Style::LabelText, m_text );
+	if( m_dimmed )
+	{
+		Imath::Color4f dimColor( 0.75, 0.75, 0.75, 1.0f );
+		style->renderText( Style::LabelText, m_text, Style::NormalState, &dimColor );
+	}
+	else
+	{
+		style->renderText( Style::LabelText, m_text );
+	}
```
Any reason we shouldn't let style make these decisions by passing `Style::DisabledState` like we're doing for the nodes and connections?

danieldresser-ie replied:
Addressed in f7d5ecf9

johnhaddon commented on src/GafferUIModule/GraphGadgetBinding.cpp:
```diff
@@ -217,6 +217,39 @@ void layoutNodes( const GraphLayout &layout, GraphGadget &graph, Gaffer::Set *no
+					const IECore::Canceller *canceller
+		)
+		{
+			std::unordered_set<const Gaffer::Plug*> activePlugs;
+			std::unordered_set<const Gaffer::Node*> activeNodes;
+
+			GraphGadget::activePlugsAndNodes( plug, context, canceller, activePlugs, activeNodes );
+
+			boost::python::list activePlugsList;
+			for( const auto &p : activePlugs )
+			{
+				activePlugsList.append( Gaffer::PlugPtr( const_cast<Plug*>( p ) ) );
+			}
+
+			boost::python::list activeNodesList;
+			for( const auto &n : activeNodes )
+			{
+				activeNodesList.append( Gaffer::NodePtr( const_cast<Node*>( n ) ) );
+			}
```
Probably need a ScopedGILRelease around this (but not `make_tuple`) for task based computations.

danieldresser-ie replied:
Addressed in 5f9175052

johnhaddon commented on src/GafferUI/StandardStyle.cpp:
```diff
@@ -1283,18 +1283,14 @@ Imath::Color3f StandardStyle::colorForState( Color c, State s, const Imath::Colo
-		// }
+		if( c == ConnectionColor )
+		{
+			result = lerp( 1.0f, 0.26f / luminance( result ), 0.85f ) * result;
```
Do we need to worry about input where `luminance( result ) < 0.26`, or worse `luminance( result ) == 0`?

danieldresser-ie replied:
`luminance( result ) < 0.26` seems fine - it would compute a multiplier > 1 which would increase the dark color to be closer than grey when disabled, which seems reasonable.

`== 0` is a problem though, since you can't multiply it up.  I should probably just toss a clamp in.

danieldresser-ie replied:
In an attempt to get folks at IE happy, I've now come up with yet another tweak on this.

I now desaturate partway, then fix the luminance to 0.31 ( slightly brighter than background ).  This seems to be a bit more consistent in making sure that different colors of noodles are visible when disabled, rather than relying on the original noodle color being different enough in luminance that it's still brighter after being blended heavily towards the background.

It works on black noodles, but is ugly looking ( it's weird that they switch from darker than background to brighter ), but that's probably OK - we're not really expecting dark noodles, as long as it's not a divide-by-zero like before?

johnhaddon replied:
It feels a bit odd to me that noodle colours would be sufficiently more important than node colours that they warrant different handling. And the black noodles thing seems less than ideal too. But it appears to work just fine for vanilla Gaffer, so if it's what we need to get this out of the door, so be it.

danieldresser-ie replied:
Current version is 9e1d125a1

johnhaddon reviewed (changes requested):
Thanks Daniel - it's gonna be great to have this over the line.

> UI style wise, I think I'm OK with this. I've tried out not desaturating the noodle colors, which I think does make things a little clearer?

I don't think it made it any worse for stock Gaffer, so if it helps out with the IE colour scheme I'm all for it.

> As for the behaviour of deactivating connections but not nodes when there is no focus node, I don't mind it visually, but the API seems pretty counterintuitive - it's not just a UI style thing, but we actually call activeForFocusNode( False ) for one and True for the other when there is no focus node. That's weird, but it's a private API, and we're probably about to rework it, so I guess we don't care that it's bad?

Yeah, it is a bit weird. I'm happy enough to back down on this one if you'd prefer.

> I was trying to finish off the final commit where I rework the actual traversal behaviour to be more correct for things like expressions on the variable name of a ContextVariables node, but for some reason, that causes crashes. I probably do need to figure out what's going on there before we merge this.

I haven't really looked at the tests or the content of `activePlugsAndNodes()`, in anticipation of them perhaps changing...

Cheers...
John

danieldresser-ie commented on include/Gaffer/Context.h:
```diff
@@ -376,6 +376,7 @@ class GAFFER_API Context : public IECore::RefCounted
 		mutable IECore::MurmurHash m_hash;
 		mutable bool m_hashValid;
 		const IECore::Canceller *m_canceller;
+		IECore::MurmurHash m_scriptContextHash;
```
Oops, removed this line in place.

danieldresser-ie commented on include/GafferUI/GraphGadget.h:
```diff
@@ -295,9 +306,44 @@ class GAFFERUI_API GraphGadget : public ContainerGadget
+		// by a mutex
+		struct PendingActive
+		{
+			std::unordered_set<const Gaffer::Plug*> activePlugs;
```
Currently, I don't think anything dereferences these pointers?  We just loop through all the Gadgets being viewed, and check if they are referencing a plug in the active list.  If some plugs have been deallocated in the meantime, it's not really a problem?

I guess in theory, there is an extremely rare possibility that there could be an active plug that gets deleted, and then a new plug get allocated at the same address, and is incorrectly labelled as active until the next update?  But that is just a small visual error in a rare case, doesn't seem too much worth worrying about?

johnhaddon replied:
Makes sense - thanks for the explanation.

johnhaddon commented on python/GafferUITest/GraphGadgetTest.py:
```diff
@@ -1406,5 +1408,336 @@ def testChangeNodeGadgetForUnviewedNode( self ) :
+			import Gaffer
+			for i in range( 50 ):
+				time.sleep( 0.2 )
+				IECore.Canceller.check( Gaffer.Context.current().canceller() )
```
Expressions should use the `context` variable rather than `Gaffer.Context.current()`. We go out of our way to provide `context` as an immutable wrapper around the true context, so shouldn't advertise this loophole (that I've only just become aware of).

danieldresser-ie replied:
Addressed in 3e0ab6d71

johnhaddon commented on python/GafferUITest/GraphGadgetTest.py:
```diff
@@ -1406,5 +1408,336 @@ def testChangeNodeGadgetForUnviewedNode( self ) :
+			args = [  s["add"]["sum"], canceller ]
+		)
+
+		startTime = time.time()
+		thread.start()
+		time.sleep( 3 )
+		canceller.cancel()
+		thread.join()
+		self.assertLess( time.time() - startTime, 5 )
```
I've become a bit wary of `sleep` and `time` comparisons for testing, because there always seems to be the odd CI host chugging so badly that it doesn't do what you want. In my recent canceller work I used condition variables to wait till something was definitely started, and had the cancellable thing loop forever unless cancelled, so we didn't need the time comparison. Might be worth doing here?

danieldresser-ie replied:
Addressed in 3e0ab6d71.

Couldn't figure out how to apply this to part of the test, so I just removed half of it.  Wasn't really testing what I wanted to test anyway.

johnhaddon commented on python/GafferUITest/GraphGadgetTest.py:
```diff
@@ -1406,5 +1408,336 @@ def testChangeNodeGadgetForUnviewedNode( self ) :
+		) )
+
+		canceller = IECore.Canceller()
+		thread = threading.Thread(
```
Doesn't really matter in the tests, but in real life it's definitely preferable to use `ParallelAlgo.runInBackgroundThread()`, so we use one of the pre-existing TBB threads. Otherwise we make a bunch of thread-local caches for the new thread, and I don't think `enumerable_thread_specific` cleans them up when the thread ends?

danieldresser-ie replied:
Addressed in 3e0ab6d.

johnhaddon reviewed (commented):
Added a few comments on the `activeWalk()` stuff and tests...

johnhaddon commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+	}
+	catch( const Gaffer::ProcessException &e )
+	{
+		// If there's an error in the graph, we can't figure out specifically which nodes contribute,
```
That seems like a pretty decent principle - certainly something we can explain easily. Let's not get into refactoring the error handling for this one :)

johnhaddon commented:
> The other thing I ran into was that my test graph used an OSL expression to connect things, and after fixing that, I got a crash, which is fixed by: 804729b

Got the above comment sent to me via a GitHub notification email, but can't find it in the web UI. Good catch on that - I'm surprised it hasn't come up already. Did a quick test here and it seems that _any_ exception (not just `IECore::Cancelled`) can cause the same crash. I think we probably need to deal with that by storing `std::current_exception()` and then rethrowing with `std::rethrow_exception()`? This one seems nasty enough that it'd be worth having a dedicated test case for it.

danieldresser-ie commented:
The abevo comment was an addendum to my reply on one of your initial comments on GraphGadget.cpp.  I can see it in the web UI, but it is collapsed by default.

Using rethrow and adding a test sounds good ... my first reaction was just to be glad it wasn't my fault.  I wonder if this fix should be a separate PR that would be back-portable - though maybe I'll end up breaking binary compatibility in order to cache the exception?

johnhaddon commented:
> I wonder if this fix should be a separate PR that would be back-portable

Sounds good - it's nasty enough that fixing for `0.59_maintenance` would be the ideal.

> though maybe I'll end up breaking binary compatibility in order to cache the exception?

I think we'd probably stash it on `RenderState`? The whole of OSLExpressionEngine is hidden anyway, so I don't think binary compatibility is an issue...

danieldresser-ie commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+			}
+			else if( const Gaffer::ContextProcessor *contextProcessorNode = runTimeCast<const Gaffer::ContextProcessor>( node ) )
+			{
+				const Gaffer::Plug *input = contextProcessorNode->inPlug()->getInput();
+				if( input )
```
I have now done the same here - though I guess it's not necessarily correct.  The question is ... if someone adds a new output plug in a subclass, does that output plug pull on the inPlug?  On the switch, NameSwitch adds the output outIndexPlug(), which does not pull on inPlug(), which means it's fine to skip tracing inPlug() when outIndexPlug is traversed.  I guess ContextVariables::combinedVariablesPlug is the same, so it is good to add this currently.  But in theory, someone could add a ContextVariables subclass that had another output that also pulled inPlug(), in which case this would now be wrong ... maybe we just hope no one does that?

danieldresser-ie replied:
Addresed in ae63d29

johnhaddon replied:
Hmm, good point. So it seems that we should fall back to the "visit all inputs" approach for any _output_ plug we don't recognise? I suspect the future work we have in mind will mean refactoring slightly to make `activeWalk()` plug-centric rather than connection-centric, so let's deal with it then. This is working for all the cases we know about, and we need to get it into people's hands.

danieldresser-ie commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+
+			// The input plug of the loop is a normal input, and will be handled below
+
+			// The next plug of the loop is evaluated in many different contexts - instead of evaluating all
+			// of them, pass a null context which will trigger a conservative evaluation of all possible inputs
```
Yeah, maybe we will want to do that eventually, though I think it would have some knock-on effects - currently, there is almost no reasonable way to make this take a long time, which allows skipping some concerns.  If we add real support for loops, then it possible for it to take a long time, and maybe we would also want to start using multiple threads for evaluation, and add some sort of busy indicator?

johnhaddon replied:
> currently, there is almost no reasonable way to make this take a long time, which allows skipping some concerns

For some definitions of reasonable :) In practice, I'll wager that people are still merrily doing things like connecting FilterResults to `SomeNode.enabled` somewhere.

danieldresser-ie commented on python/GafferUITest/GraphGadgetTest.py:
```diff
@@ -1406,5 +1408,336 @@ def testChangeNodeGadgetForUnviewedNode( self ) :
+
+		def testThreadFunc( plug, canceller ):
+			try:
+				# TODO - should probably omit canceller from this API, and just take a context?
```
I've removed the canceller, but kept the explicit Context - it just feels to me like in most uses of this, it helps to have the context be explicit.  I can switch to `Context::current()` if you prefer.

johnhaddon replied:
I do think it's worth removing the `canceller` argument from `activeWalk()` - we are relying on it being the same as `context->canceller()` (so that any computes see the same canceller as we're checking in `activeWalk()` itself), so having it as an extra argument just gives us a chance to get it wrong.

danieldresser-ie commented on include/GafferUI/GraphGadget.h:
```diff
@@ -295,9 +306,44 @@ class GAFFERUI_API GraphGadget : public ContainerGadget
 		std::vector<float> m_dragSnapOffsets[2]; // offsets in x and y
 		std::vector<Imath::V2f> m_dragSnapPoints; // specific points that are also target for point snapping
 		int m_dragMergeGroupId;
+		IECore::MurmurHash m_scriptContextHash;
```
Addressed in 078a3777

danieldresser-ie commented on include/GafferUI/GraphGadget.h:
```diff
@@ -295,9 +306,44 @@ class GAFFERUI_API GraphGadget : public ContainerGadget
+		// Track if we need to run an active state update.  If true, then if there isn't already an
+		// m_activeTask running, we need to start one.  ( Helpful to track separately so we know
+		// we need to restart if the task gets cancelled somehow )
+		mutable std::atomic_bool m_activeStateDirty;
```
Addressed in 078a377

danieldresser-ie commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+					// We are handling just the value inputs specially
+					for( auto &inSourcePlug : Gaffer::Plug::InputRange( *nameSwitchNode->inPlugs() ) )
+					{
+						const Gaffer::NameValuePlug *inputNameValue = runTimeCast<const Gaffer::NameValuePlug>( inSourcePlug.get() );
```
Addresed in ae63d291

danieldresser-ie commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+		for( const Gaffer::Plug* handledPlug : handledBySpecialCase )
+		{
+			// TODO - using isAncestor here is pretty ugly, alternative would be doing recursive
+			// traversal ourselves so we could terminate if we encounter a plug that is handled?
```
Addresed in ae63d29

danieldresser-ie commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+			{
+				// Parents are connected, don't need to trace individual children
+				continue;
+			}
```
Addresed in ae63d29.  I did note while working on this that current tests are just on FloatPlug rather than a hierarchical plug like ScenePlug, so this code is not tested well by the unit tests.  Maybe need to add some more.

danieldresser-ie commented on python/GafferUITest/GraphGadgetTest.py:
```diff
@@ -1406,5 +1408,336 @@ def testChangeNodeGadgetForUnviewedNode( self ) :
+			try:
+				# TODO - should probably omit canceller from this API, and just take a context?
+				GafferUI.GraphGadget._activePlugsAndNodes( plug, Gaffer.Context( Gaffer.Context(), canceller ), canceller )
+			except:
```
Addressed in 3e0ab6d71

danieldresser-ie commented:
I think I've now addressed all feedback.  Also added c196860c which seems like a definite benefit, and e72064b75 which sets the TaskPlug color to what IE wants ( though that can certainly be moved to an IE config if you prefer ).

Planning to let you review before squashing.  There are some whitespace errors the CI picked up - I'll fix those while squashing.

johnhaddon commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -763,14 +764,27 @@ float StandardNodeGadget::focusBorderWidth() const
+void StandardNodeGadget::setHighlighted( bool highlighted )
+{
+	NodeGadget::setHighlighted( highlighted );
+	m_highlightedStandardNodeGadget = highlighted;  // Don't really get why the base class one is protected
```
Why do we need to duplicate the state? I think we can just call `getHighlighted()` when we want to access it can't we?

johnhaddon reviewed (commented):
Thanks Daniel - looking good. I've made one comment above `setHighlighted()` not needing to duplicate state, and there's a comment on the previous review about removing `canceller` from `activeWalk()`. If you could deal with those while squashing that would be grand...

danieldresser-ie commented:
Squashed all previous commits, and fixed whitespace.  Addressed the silly thing I did with missing getHighlighted() in place.

As a final step, I did add some tests for compound plugs, which did actually turn up a corner case I wasn't handling ( which is rare but possible in production, with a Switch or ContextProcessor with a ScenePlug input where just the bound is connected to something ).  Re-arranged a bit to address this, feels like I've centralized some logic a bit better too - hopefully makes sense to you.

Anyway, check out the last 3 commits before I squash.

danieldresser-ie commented:
Squashed and updated changes.  If the tests pass again, then maybe we blame that one Mac test on gremlins, and merge?

johnhaddon commented:
> Squashed and updated changes. If the tests pass again, then maybe we blame that one Mac test on gremlins, and merge?

That's my inclination at the moment. I ran the tests some more on my Mac here and couldn't get anything to fail...
